src/NHibernate/Engine/ActionQueue.cs

```diff
@@ -195,8 +205,16 @@ private void InnerExecute(IExecutable executable)
 
 		private void RegisterCleanupActions(IExecutable executable)
 		{
-			beforeTransactionProcesses.Register(executable.BeforeTransactionCompletionProcess);
-			afterTransactionProcesses.Register(executable.AfterTransactionCompletionProcess);
+			if (executable is IAsyncExecutable asyncExecutable)
+			{
+				beforeTransactionProcesses.Register(asyncExecutable.BeforeTransactionCompletionProcess);
+				afterTransactionProcesses.Register(asyncExecutable.AfterTransactionCompletionProcess);
```

Comment:
It could be:

```c#
RegisterProcess(asyncExecutable.BeforeTransactionCompletionProcess);
RegisterProcess(asyncExecutable.AfterTransactionCompletionProcess);
```

Slightly less code written, and the compiler will likely inline the call anyway.